go/libraries/doltcore/sqle/dolt_foreign_key_status_table_function.go

```diff
@@ -0,0 +1,342 @@
+		if err != nil {
+			return false, fmt.Errorf("unable to get table %s: %w", fk.TableName, err)
+		}
+		tableSch, err := table.GetSchema(ctx)
+		if err != nil {
+			return false, fmt.Errorf("unable to get schema for table %s: %w", fk.TableName, err)
+		}
+		refTable, _, err := root.GetTable(ctx, fk.ReferencedTableName)
+		if err != nil {
+			return false, fmt.Errorf("unable to get table %s: %w", fk.ReferencedTableName, err)
+		}
+		refTableSch, err := refTable.GetSchema(ctx)
+		if err != nil {
+			return false, fmt.Errorf("unable to get schema for table %s: %w", fk.ReferencedTableName, err)
+		}
+
+		rows, err := getForeignKeyRows(fk, tableSch, refTableSch)
+		if err != nil {
+			return false, err
+		}
+		fkRows = append(fkRows, rows...)
+		return true, nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("unable to iterate foreign keys: %w", err)
+	}
+
+	iter := &foreignKeyStatusTableFunctionRowIter{
+		fkRows: fkRows,
+		index:  0,
+	}
+	return iter, nil
+}
+
+func getForeignKeyRows(fk doltdb.ForeignKey, tableSch schema.Schema, refTableSch schema.Schema) ([]sql.Row, error) {
+
+	var tableColumnNames []string
+	for _, colTag := range fk.TableColumns {
+		col, ok := tableSch.GetAllCols().GetByTag(colTag)
+		if !ok {
+			return nil, fmt.Errorf("unable to get column for tag %d", colTag)
+		}
+		tableColumnNames = append(tableColumnNames, col.Name)
+	}
+	var referenceTableColumNames []string
+	for _, colTag := range fk.ReferencedTableColumns {
+		col, ok := refTableSch.GetAllCols().GetByTag(colTag)
+		if !ok {
+			return nil, fmt.Errorf("unable to get column for tag %d", colTag)
+		}
+		referenceTableColumNames = append(referenceTableColumNames, col.Name)
+	}
+
+	baseRow := sql.NewRow(
+		fk.Name,                 // 0
+		fk.IsResolved(),         // 1
+		fk.TableName,            // 2
+		fk.TableIndex,           // 3
+		"",                      // 4
+		fk.ReferencedTableName,  // 5
+		fk.ReferencedTableIndex, // 6
+		"",                      // 7
+		fk.OnUpdate.String(),    // 8
+		fk.OnDelete.String(),    // 9
+		"",                      // 10
+		"",                      // 11
+	)
+	rows := []sql.Row{baseRow}
+	if fk.IsResolved() {
+		if len(tableColumnNames) > 0 {
+			rows = addColToAllRows(rows, 4, tableColumnNames...)
+		}
+		if len(referenceTableColumNames) > 0 {
+			rows = addColToAllRows(rows, 7, referenceTableColumNames...)
+		}
+	} else {
+		if len(fk.UnresolvedFKDetails.TableColumns) > 0 {
+			rows = addColToAllRows(rows, 10, fk.UnresolvedFKDetails.TableColumns...)
+		}
+		if len(fk.UnresolvedFKDetails.ReferencedTableColumns) > 0 {
+			rows = addColToAllRows(rows, 11, fk.UnresolvedFKDetails.ReferencedTableColumns...)
+		}
+	}
+
+	return rows, nil
+}
+
+func addColToAllRows(rows []sql.Row, colIndex int, data ...string) []sql.Row {
+	newRows := []sql.Row{}
+	for _, datum := range data {
+		for _, row := range rows {
+			newRow := row.Copy()
+			newRow[colIndex] = datum
+			newRows = append(newRows, newRow)
```

Comment:
I don't think you need to make a new row array each time;  just pass in a pointer to the row array and modifying the rows.

Reply:
We do want to make copies of rows.
The idea here is that we need to return a list of fields for one of the columns.
We're doing this for multiple columns, but let's just consider `table_column`. So if we have 2 values for this column, "foo" and "bar", we want to produce 2 almost-identical  rows of values: (a,b,c,foo,d) and (a,b,c,bar,d).